[MPS] Add CompileShader method

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141478

This allows one to do something like that
```python
import torch
x = torch.ones(10, device="mps")
m = torch.mps._compile_shader("""
   kernel void foo(device float* x, uint idx [[thread_position_in_grid]]) {
     x[idx] += idx;
   }
")
m.foo(x)
```

And in general enables writing custom operators using Metal shaders purely in Python

malfet committed:
[WIP] Add CompileShader method

[ghstack-poisoned]

malfet committed:
Update on "[WIP] Add CompileShader method"




[ghstack-poisoned]

malfet committed:
Update on "[WIP] Add CompileShader method"




[ghstack-poisoned]

malfet committed:
Update on "[WIP] Add CompileShader method"




[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

qqaatw commented:
I'm curious about the use of this addition since it accepts one argument only. I assume there is no cuda counterpart in PyTorch.

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet commented:
> I'm curious about the use of this addition since it accepts one argument only. I assume there is no cuda counterpart in PyTorch.

@qqaatw it's still work in progress :) But final version will accept any number of arguments. As for usefulness, at the very least it unblocks quick prototyping and usecases like shader toys, but in the long run it can enable a pretty advanced codegen.

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add CompileShader method"


This allows one to do something like that
```python
import torch
x = torch.ones(10, device="mps")
m = torch.mps._compile_shader("""
   kernel void foo(device float* x, uint idx [[thread_position_in_grid]]) {
     x[idx] += idx;
   }
")
m.foo(x)
```

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet commented on test/test_mps.py:
```diff
@@ -12480,6 +12480,39 @@ def ones(device):
             cpu_tensor = ones("cpu")
             self.assertEqual(mps_tensor.cpu(), cpu_tensor)
 
+class TestMetalLibrary(TestCaseMPS):
```
Done

malfet commented on torch/mps/__init__.py:
```diff
@@ -140,6 +140,12 @@ def recommended_max_memory() -> int:
     return torch._C._mps_recommendedMaxMemory()
 
 
+def _compile_shader(source: str):
```
I think this should be done as a separate wrapper in the Python layer, where we create an abstraction (similar to TensorIterator) and developer only need to implement the function that is executed in the main loop

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint + MPS are green"

malfet commented:
@pytorchbot revert -m "Broke doctests, by trying to run MPS example on Linux" -c nosignal

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "No pain no gain"
